### PR TITLE
Feature 4 : Make Todo done functionality implemented.

### PR DESCRIPTION
**Pull request by saa-wtag**
[Task-4: Make Todo Done](https://trello.com/c/SBAAcrhU)
This PR introduces the functionality of Mark a todo by clicking the done button beside edit button. By clicking done one can make the todo title strike through. In edit mode, users can make the todo mark done, without extra updating the title.

**Functionalities:**
There is a new button beside Edit and Delete named Done.
If Done is clicked, the text will have a strike through style.
When a Todo is marked Done, the Edit & Done button will disappear.

Please have a close look on this PR and kindly put your valuable comments.

https://github.com/sad-wtag/ToDoApp_VanillaJS/assets/164306406/c3048d92-f1b4-45a7-a5bd-eefe41e83277

### CODE REVIEW

File: js/script.js [hunk 1]
```diff
@@ -44,17 +57,22 @@ const createTask = (taskTitle) => {
...
 
     if (task.editMode) {
       const $inputField = document.createElement("input");
       $inputField.type = "text";
       $inputField.value = task.title;
 
-      const $updateButton = createButton("Update", () => updateHandler(task, $inputField.value));
+      const $updateButton = createButton("Update", () =>
+        updateHandler(task, $inputField.value)
+      );
```
Comment (rza-wtag) [inline js/script.js]: It's not part of this specific PR I believe

File: js/script.js [hunk 2]
```diff
@@ -63,23 +81,29 @@ const renderTasks = () => {
 
       $tasksList.append($inputField, $updateButton, $cancelButton);
     } else {
-      const $deleteButton = createButton("Delete", () => deleteHandler(task.id));
+      const $deleteButton = createButton("Delete", () =>
+        deleteHandler(task.id)
+      );
```
Comment (rza-wtag) [inline js/script.js]: ditto

File: js/script.js [hunk 3]
```diff
@@ -63,23 +81,29 @@ const renderTasks = () => {
...
+      }
     }
 
     $taskList.appendChild($tasksList);
   });
 };
 
 const cancelEdit = () => {
-  tasks.forEach(task => task.editMode = false); 
+  tasks.forEach((task) => (task.editMode = false));
```
Comment (rza-wtag) [inline js/script.js]: why this change is here.?

File: js/utilities.js [hunk 4]
```diff
@@ -9,15 +9,15 @@ export const showToastMessage = (message) => {
...
-    const trimmedValue = $inputField.value.trim();
-    $updateButton.disabled = !(
-      trimmedValue.length > 0 && trimmedValue !== currentTask.title
-    );
-  };
+  const trimmedValue = $inputField.value.trim();
+  $updateButton.disabled = !(
+    trimmedValue.length > 0 && trimmedValue !== currentTask.title
+  );
+};
```
Comment (rza-wtag) [inline js/utilities.js]: Please follow the task description of the Trello board.

File: js/utilities.js [hunk 5]
```diff
@@ -9,15 +9,15 @@ export const showToastMessage = (message) => {
...
-    $button.addEventListener("click", onClick);
-    return $button;
-  };
\ No newline at end of file
+export const createButton = (text, onClick) => {
+  const $button = document.createElement("button");
+  $button.innerText = text;
+  $button.addEventListener("click", onClick);
+  return $button;
+};
```
Comment (rza-wtag) [inline js/utilities.js]: ditto

**Review by sal-wtag — COMMENTED** [review]
@saa-wtag bhai, please resolve the pending reviews. Thank you.

File: js/script.js (hunk 1 above)
Comment (saa-wtag) [reply js/script.js]: Fixed, bhai.

File: js/script.js (hunk 3 above)
Comment (saa-wtag) [reply js/script.js]: solved, bhai.

File: js/script.js (hunk 2 above)
Comment (saa-wtag) [reply js/script.js]: solved, bhai.

File: js/utilities.js (hunk 4 above)
Comment (saa-wtag) [reply js/utilities.js]: I didn't get, what you intended to make me understand, bhai !

Comment (saa-wtag) [reply js/utilities.js]: solved

File: js/utilities.js (hunk 5 above)
Comment (saa-wtag) [reply js/utilities.js]: It is required to make the codebase modularized.

**Comment by saa-wtag** [comment]
> @saa-wtag bhai, please resolve the pending reviews. Thank you.

done bhai

**Review by sal-wtag — APPROVED** [review]

File: js/script.js [hunk 6]
```diff
@@ -38,6 +38,14 @@ const updateHandler = (task, newTitle) => {
...
 };
 
+const doneHandler = (taskId) => {
+  const task = tasks.find((task) => task.id === taskId);
+  if (task) {
+    task.done = !task.done;
+    renderTasks();
+  }
+};
+
```
Comment (maa-wtag) [inline js/script.js]: @saa-wtag  from the function statement it seems like we are toggling the task from Done to Not Done and vice versa, I think that is not the requirement in the story itself.

File: js/script.js [hunk 7]
```diff
@@ -57,6 +65,9 @@ const renderTasks = () => {
     const $tasksList = document.createElement("li");
     const $titleElement = document.createElement("span");
     $titleElement.textContent = task.title;
+    if (task.done) {
+      $titleElement.style.textDecoration = "line-through";
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
      $taskTitle.style.textDecoration = "line-through";
```

File: js/script.js [hunk 8]
```diff
@@ -76,8 +87,14 @@ const renderTasks = () => {
       const $editButton = createElement("Edit", "button", () =>
         editHandler(task)
       );
+      const $doneButton = createElement("Done", "button", () =>
+        doneHandler(task.id)
+      );
 
-      $tasksList.append($titleElement, $deleteButton, $editButton);
+      $tasksList.append($titleElement, $deleteButton);
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
      $tasksList.append($taskTitle, $deleteButton);
```

**Review by maa-wtag — CHANGES_REQUESTED** [review]

File: js/script.js [hunk 9]
```diff
@@ -38,6 +38,14 @@ const updateTask = (task, newTitle) => {
   renderTasks();
 };
 
+const doneTask = (taskId) => {
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
const completeTask = (taskId) => {
```

File: js/script.js [hunk 10]
```diff
@@ -38,6 +38,14 @@ const updateTask = (task, newTitle) => {
...
+  const task = tasks.find((task) => task.id === taskId);
+  if (task) {
+    task.done = !task.done;
```
Comment (maa-wtag) [inline js/script.js]: still seems to toggle the Done state, is it what the story requirement says?

File: js/script.js [hunk 11]
```diff
@@ -58,7 +66,8 @@ const renderTasks = () => {
       deleteTask,
       editTask,
       updateTask,
-      cancelEdit
+      cancelEdit,
+      doneTask
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
      completeTask
```

File: js/utilities.js [hunk 12]
```diff
@@ -27,12 +27,16 @@ export const createTaskElement = (
   deleteTask,
   editTask,
   updateTask,
-  cancelEdit
+  cancelEdit,
+  doneTask
```
Comment (maa-wtag) [inline js/utilities.js]:
```suggestion
  completeTask
```

File: js/utilities.js [hunk 13]
```diff
@@ -43,16 +47,24 @@ export const createTaskElement = (
...
+    });
 
     $taskItem.append($inputField, $updateButton, $cancelButton);
   } else {
     const $deleteButton = createElement("Delete", "button", () =>
       deleteTask(task.id)
     );
     const $editButton = createElement("Edit", "button", () => editTask(task));
+    const $doneButton = createElement("Done", "button", () =>
+      doneTask(task.id)
```
Comment (maa-wtag) [inline js/utilities.js]:
```suggestion
      completeTask(task.id)
```

File: js/script.js (hunk 6 above)
Comment (saa-wtag) [reply js/script.js]: Done, bhai.

File: js/script.js (hunk 10 above)
Comment (saa-wtag) [reply js/script.js]: Bhai, sorry for overlook the issue, now it is resolved.

File: js/script.js (hunk 11 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: js/script.js [hunk 14]
```diff
@@ -38,6 +38,14 @@ const updateTask = (task, newTitle) => {
   renderTasks();
 };
 
+const completeTask = (taskId) => {
+  const task = tasks.find((task) => task.id === taskId);
+  if (task) {
```
Comment (maa-wtag) [inline js/script.js]:
```suggestion
  if (task && !task.done) {
```

File: js/utilities.js [hunk 15]
```diff
@@ -27,12 +27,16 @@ export const createTaskElement = (
...
   updateTask,
-  cancelEdit
+  cancelEdit,
+  completeTask
 ) => {
   const $taskItem = document.createElement("li");
 
   const $titleElement = document.createElement("span");
   $titleElement.textContent = task.title;
+  if (task.done) {
```
Comment (maa-wtag) [inline js/utilities.js]:
```suggestion
  if (task && task.done) {
```

**Review by maa-wtag — APPROVED** [review]

File: js/script.js [hunk 16]
```diff
@@ -37,6 +37,15 @@ const updateTask = (task, newTitle) => {
...
 };
 
+const completeTask = (taskId) => {
+  const task = tasks.find((task) => task.id === taskId);
+  if (task && !task.done) {
+    task.done = true;
+    task.isEditing = false;
+    renderTasks();
+  }
+};
```
Comment (hhu-wtag) [inline js/script.js]: Generally it's a good practice to reduce branches in code whenever we can.

```suggestion
const completeTask = (taskId) => {
  const task = tasks.find((task) => task.id === taskId);
  
  if (task === undefined){
    return
  }
  
  task.done = true;
  task.isEditing = false;
  
  renderTasks();
};
```

File: js/utilities.js
Comment (hhu-wtag) [inline js/utilities.js:1]: Let's create an elements.js file and move the logics of element creations to the file.

**Review by hhu-wtag — APPROVED** [review]

File: js/script.js (hunk 16 above)
Comment (saa-wtag) [reply js/script.js]: Done

File: js/utilities.js
Comment (saa-wtag) [reply js/utilities.js:1]: Bhai, from the elements.js file i am currently exporting DOM elements  to othe js files. Should I create elements also here besides exporting DOM elements ?
